File: backend/app/services/google_slides_parser.py
from typing import Optional, List
from app.config.google_slides import get_google_slides_client


def fetch_presentation(presentation_id: str):
    """Google SlidesのプレゼンテーションIDからスライドデータを取得"""
    slides = get_google_slides_client()
    
    try:
        presentation = slides.presentations().get(presentationId=presentation_id).execute()
        return presentation
    except Exception as e:
        print(f"Error fetching presentation: {e}")
        raise ValueError(f"Failed to fetch presentation: {str(e)}")


class SlideElement:
    def __init__(
        self,
        element_id: str,
        element_type: str,
        content: Optional[str] = None,
        position: Optional[dict] = None,
        style: Optional[dict] = None,
        image_url: Optional[str] = None,
        children: Optional[List["SlideElement"]] = None,
    ):
        self.id = element_id
        self.type = element_type
        self.content = content
        self.position = position
        self.style = style
        self.image_url = image_url
        self.children = children or []


def extract_slide_elements(page: dict) -> List[SlideElement]:
    """スライドページから要素を抽出"""
    elements: List[SlideElement] = []
    
    page_elements = page.get("pageElements", [])
    for element in page_elements:
        extracted = extract_element(element)
        if extracted:
            elements.append(extracted)
    
    return elements


def extract_element(element: dict) -> Optional[SlideElement]:
    """単一要素を抽出"""
    element_id = element.get("objectId")
    if not element_id:
        return None
    
    # 位置情報を抽出
    transform = element.get("transform", {})
    position = None
    if transform:
        translate_x = transform.get("translateX", {}).get("magnitude", 0) / 9525  # EMU to pixels
        translate_y = transform.get("translateY", {}).get("magnitude", 0) / 9525
        width = transform.get("width", {}).get("magnitude", 0) / 9525
        height = transform.get("height", {}).get("magnitude", 0) / 9525
        position = {
            "x": translate_x,
            "y": translate_y,
            "width": width,
            "height": height,
        }
    
    base_element = SlideElement(
        element_id=element_id,
        element_type="shape",
        position=position,
    )
    
    # テキスト要素
    shape = element.get("shape", {})
    if shape.get("shapeType") == "TEXT_BOX" or shape.get("text"):
        text_content = extract_text_content(shape.get("text", {}))
        return SlideElement(
            element_id=element_id,
            element_type="text",
            content=text_content["text"],
            style=text_content.get("style"),
            position=position,
        )
    
    # 画像要素
    if "image" in element:
        image = element["image"]
        return SlideElement(
            element_id=element_id,
            element_type="image",
            image_url=image.get("contentUrl") or image.get("sourceUrl"),
            position=position,
        )
    
    # 図形要素
    if shape:
        shape_type = shape.get("shapeType")
        if shape_type and shape_type != "TEXT_BOX":
            return SlideElement(
                element_id=element_id,
                element_type="shape",
                content=shape_type,
                position=position,
            )
    
    # テーブル要素
    if "table" in element:
        table = element["table"]
        return SlideElement(
            element_id=element_id,
            element_type="table",
            content=extract_table_content(table),
            position=position,
        )
    
    # グループ要素
    if "elementGroup" in element:
        element_group = element["elementGroup"]
        children = []
        for child in element_group.get("children", []):
            child_element = extract_element(child)
            if child_element:
                children.append(child_element)
        
        return SlideElement(
            element_id=element_id,
            element_type="group",
            children=children,
            position=position,
        )
    
    return base_element


def extract_text_content(text_element: dict) -> dict:
    """テキストコンテンツを抽出"""
    text_elements = text_element.get("textElements", [])
    if not text_elements:
        return {"text": ""}
    
    full_text = ""
    style = None
    
    for text_el in text_elements:
        text_run = text_el.get("textRun", {})
        if text_run:
            full_text += text_run.get("content", "")
            
            text_style = text_run.get("textStyle", {})
            if text_style:
                style = {
                    "fontSize": text_style.get("fontSize", {}).get("magnitude", 0) / 100 if text_style.get("fontSize") else None,
                    "fontFamily": text_style.get("fontFamily"),
                    "color": rgb_to_hex(text_style.get("foregroundColor", {}).get("opaqueColor", {}).get("rgbColor")),
                    "bold": text_style.get("bold"),
                    "italic": text_style.get("italic"),
                }
    
    return {"text": full_text.strip(), "style": style}


def extract_table_content(table: dict) -> str:
    """テーブルコンテンツを抽出"""
    table_rows = table.get("tableRows", [])
    if not table_rows:
        return ""
    
    rows = []
    for row in table_rows:
        table_cells = row.get("tableCells", [])
        cells = []
        for cell in table_cells:
            text_content = extract_text_content(cell.get("text", {}).get("textContent", {}))
            cells.append(text_content["text"] or "")
        rows.append(" | ".join(cells))
    
    return "\n".join(rows)


def rgb_to_hex(rgb: Optional[dict]) -> Optional[str]:
    """RGBカラーをHEXに変換"""
    if not rgb:
        return None
    
    r = round((rgb.get("red", 0) or 0) * 255)
    g = round((rgb.get("green", 0) or 0) * 255)
    b = round((rgb.get("blue", 0) or 0) * 255)
    
    return f"#{r:02x}{g:02x}{b:02x}"

File: backend/src/services/google-slides-parser.ts
import { getGoogleSlidesClient } from "../config/google-slides.js"
import type { slides_v1 } from "googleapis"

/**
 * Google SlidesのプレゼンテーションIDからスライドデータを取得
 */
export async function fetchPresentation(presentationId: string): Promise<slides_v1.Schema$Presentation> {
  const slides = getGoogleSlidesClient()
  
  try {
    const presentation = await slides.presentations.get({
      presentationId,
    })
    
    if (!presentation.data) {
      throw new Error("Presentation data is null")
    }
    
    return presentation.data
  } catch (error) {
    console.error("Error fetching presentation:", error)
    throw new Error(`Failed to fetch presentation: ${error instanceof Error ? error.message : "Unknown error"}`)
  }
}

/**
 * スライドオブジェクトから詳細な構造を抽出
 */
export interface SlideElement {
  id: string
  type: "text" | "image" | "shape" | "table" | "video" | "line" | "group"
  content?: string
  position?: {
    x: number
    y: number
    width: number
    height: number
  }
  style?: {
    fontSize?: number
    fontFamily?: string
    color?: string
    bold?: boolean
    italic?: boolean
    alignment?: string
  }
  imageUrl?: string
  children?: SlideElement[]
}

/**
 * スライドページから要素を抽出
 */
export function extractSlideElements(
  page: slides_v1.Schema$Page
): SlideElement[] {
  const elements: SlideElement[] = []

  if (!page.pageElements) {
    return elements
  }

  for (const element of page.pageElements) {
    const extracted = extractElement(element)
    if (extracted) {
      elements.push(extracted)
    }
  }

  return elements
}

/**
 * 単一要素を抽出
 */
function extractElement(
  element: slides_v1.Schema$PageElement
): SlideElement | null {
  if (!element.objectId) {
    return null
  }

  const baseElement: SlideElement = {
    id: element.objectId,
    type: "shape",
    position: element.transform
      ? {
          x: (element.transform.translateX?.magnitude || 0) / 9525, // EMU to pixels (1 EMU = 1/9525 cm)
          y: (element.transform.translateY?.magnitude || 0) / 9525,
          width: (element.transform.width?.magnitude || 0) / 9525,
          height: (element.transform.height?.magnitude || 0) / 9525,
        }
      : undefined,
  }

  // テキスト要素
  if (element.shape?.shapeType === "TEXT_BOX" || element.shape?.text) {
    const textContent = extractTextContent(element.shape.text)
    return {
      ...baseElement,
      type: "text",
      content: textContent.text,
      style: textContent.style,
    }
  }

  // 画像要素
  if (element.image) {
    return {
      ...baseElement,
      type: "image",
      imageUrl: element.image.contentUrl || element.image.sourceUrl || undefined,
    }
  }

  // 図形要素
  if (element.shape) {
    const shapeType = element.shape.shapeType
    if (shapeType && shapeType !== "TEXT_BOX") {
      return {
        ...baseElement,
        type: "shape",
        content: shapeType,
      }
    }
  }

  // テーブル要素
  if (element.table) {
    return {
      ...baseElement,
      type: "table",
      content: extractTableContent(element.table),
    }
  }

  // グループ要素
  if (element.elementGroup) {
    const children = element.elementGroup.children
      ?.map((child) => extractElement(child))
      .filter((e): e is SlideElement => e !== null) || []

    return {
      ...baseElement,
      type: "group",
      children,
    }
  }

  return baseElement
}

/**
 * テキストコンテンツを抽出
 */
function extractTextContent(
  textElement?: slides_v1.Schema$TextContent
): { text: string; style?: SlideElement["style"] } {
  if (!textElement?.textElements) {
    return { text: "" }
  }

  let fullText = ""
  let style: SlideElement["style"] | undefined

  for (const textEl of textElement.textElements) {
    if (textEl.textRun) {
      fullText += textEl.textRun.content || ""
      
      if (textEl.textRun.textStyle) {
        const ts = textEl.textRun.textStyle
        style = {
          fontSize: ts.fontSize?.magnitude ? ts.fontSize.magnitude / 100 : undefined, // ポイントからピクセルに変換
          fontFamily: ts.fontFamily || undefined,
          color: ts.foregroundColor?.opaqueColor?.rgbColor
            ? rgbToHex(ts.foregroundColor.opaqueColor.rgbColor)
            : undefined,
          bold: ts.bold || undefined,
          italic: ts.italic || undefined,
        }
      }
    }
  }

  return { text: fullText.trim(), style }
}

/**
 * テーブルコンテンツを抽出
 */
function extractTableContent(table: slides_v1.Schema$Table): string {
  if (!table.tableRows) {
    return ""
  }

  const rows = table.tableRows.map((row) => {
    if (!row.tableCells) return ""
    return row.tableCells
      .map((cell) => {
        const text = extractTextContent(cell.text?.textContent).text
        return text || ""
      })
      .join(" | ")
  })

  return rows.join("\n")
}

/**
 * RGBカラーをHEXに変換
 */
function rgbToHex(
  rgb: slides_v1.Schema$RgbColor | undefined
): string | undefined {
  if (!rgb) return undefined

  const r = Math.round((rgb.red || 0) * 255)
  const g = Math.round((rgb.green || 0) * 255)
  const b = Math.round((rgb.blue || 0) * 255)

  return `#${[r, g, b].map((x) => x.toString(16).padStart(2, "0")).join("")}`
}
